content/learn/intermediate/database.md

```diff
@@ -0,0 +1,399 @@
+
+As an example, to find the rank of the entry with the key `test` for the current user in all of this user's entries, you
+may want to use the following code:
+
+```go
+{{$rank := dbRank (sdict "userID" .User.ID) .User.ID "test"}}
+The specified entry's rank is {{$rank}}.
+```
+
+Note that this function **returns** the rank, so to avoid random spam, don't forget to assign it to a variable.
+
+## Appendix
+
+### Patterns
+
+As mentioned earlier, we use patterns for a set of functions. Obviously, you need to know what they are and how to use
+them. The patterns are based on PostgreSQL patterns, so if you're familiar with them, you're good to go. If not, don't
+worry, they're quite easy to understand.
+
+There's only two special characters you need to know: `%` and `_`. More are not used, just those two.
+In case you need to escape those, prepend them with a backslash `\`. Their respective purpose is also quite simple:
+
+- The percent sign `%` matches any sequence of zero or more characters
+- The underscore `_` matches any single character
+
+Okay, with that in mind, let's take a look at an example. The following pattern will match anything that starts with
+the letter `l` and ends in `n`.
+
+```text
+l%n
+```
+
+The following example showcases the usage of `_`.
+
+```text
+hel_o
+```
+
+This pattern matches words such as `hello`, `helgo`, `heloo`.
+
+### Serialization
+
+Saving values with custom types to database may result in their values being _serialized_ to a different type, meaning
+that you might have to convert it back to its original type when retrieving. For example, saving the result of a `cembed`
+to call to database will result in it becoming a `map[string] interface{}`.  The following code will showcase this
+behavior:
+
+```go
+{{$embed := cembed "description" "Serialization!"}}
+{{printf "Type before storing: %T" $embed}}
+{{dbSet .User.ID "serialization_example" $embed}}
+{{$embed_retrieved := (dbGet .User.ID "serialization_example").Value}}
+{{printf "Type after retrieving, before converting: %T" $embed_retrieved}}
+{{printf "Type afer retrieving, after converting: %T" (cembed $embed_retrieved)}}
+{{dbDel .User.ID "serialization_example"}}
+```
+
+However, most commonly used types will be saved with their type information intact, meaning that there will be no need
+to convert them after retrieval. In particular, `sdict`, `dict`, and `cslice` may be saved directly to database and will
+retain their original types.
+
+### Storing IDs
+
+You might have noticed that, whenever you're storing a user ID, channel ID, etc. into your database, it will come back
+as a weird value, such as `5.241379415938826e+17`. This is because they're saved as floats, hence it being messed up.
+To prevent this, simply convert them to a string before storing and converting back to `int` upon retrieving, like so:
+
+```go
+{{dbSet 2000 "someKey" (str .User.ID)}}
+{{$userID_received := toInt (dbGet 2000 "someKey").Value}}
+{{dbDel 2000 "someKey"}}
+```
```

Comment:
I think you meant to try and remove `str` (`toString`), rather than `toInt`?

Also, I'm not sure where GitHub messed up with parsing for highlighting purposes, but if I remove that single backtick, the page won't render correctly anymore; highlighting is proper in my editor.

Reply:
Yes, you're correct on both counts.
Github was ending the suggestion code block there and it looked quite messy (though I could have omitted that entirely now that I reread)